### PR TITLE
KAFKA-12380 Executor in Connect's Worker is not shut down when the worker is

### PR DESCRIPTION
**Pull request by karuturi**
added shutdown for executors and tests for the same

### Committer Checklist (excluded from commit message)
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by karuturi** [comment]
I could not find the exact failed testcases. I can see assertion errors in StreamsUncaughtExceptionHandlerIntegrationTest. If I run it locally, it's taking forever. 

Can someone help me find the failed tests?
The change I added is very isolated. Should not cause any failures ideally.

**Comment by karuturi** [comment]
I downloaded the console log and a grep showed below tests

cat consoleText.txt| grep FAILED
[2022-03-28T14:02:18.545Z] SocketServerTest > testIdleConnection() FAILED
[2022-03-28T14:44:11.437Z] TransactionsBounceTest > testWithGroupId() FAILED
[2022-03-28T14:59:52.526Z] DynamicConnectionQuotaTest > testDynamicListenerConnectionCreationRateQuota() FAILED
[2022-03-28T15:04:10.646Z] KRaftClusterTest > testLegacyAlterConfigs() FAILED

all of them are successful if ran individually

**Comment by showuon** [comment]
@karuturi , thanks for the PR. But I found there was a long pending PR that is addressing the same issue (https://github.com/apache/kafka/pull/10337). Could we wait for a few days to see if the author wants to continue the old PR or not, before we start review this PR? (I've left a comment in that PR). 

Thanks.

**Comment by karuturi** [comment]
@showuon Sure, can wait. 
Hope we will merge one of them soon :)

**Comment by karuturi** [comment]
Hi @showuon
As there is no response on the other PR, can we take this forward?

**Comment by karuturi** [comment]
@showuon Can we merge this please?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 1]
```diff
@@ -222,6 +223,14 @@ public void stop() {
...
 
         workerConfigTransformer.close();
+        executor.shutdown();
+        try {
+            if (!executor.awaitTermination(EXECUTOR_SHUTDOWN_TERMINATION_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+                executor.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            executor.shutdownNow();
+        }
```
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Not sure if we should follow the 2-phase shutdown in "usage example" in java doc?
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ExecutorService.html

**Review by showuon — COMMENTED** [review]
LGTM! Left one comment. And thanks for the tests.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 1 above)
Comment (karuturi) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: I missed preserving interrupt status if `InterruptedException` occurs. Also added the log as suggested in java doc.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java [hunk 2]
```diff
@@ -222,6 +223,20 @@ public void stop() {
...
 
         workerConfigTransformer.close();
+        executor.shutdown();
+        try {
+            // Wait a while for existing tasks to terminate
+            if (!executor.awaitTermination(EXECUTOR_SHUTDOWN_TERMINATION_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+                executor.shutdownNow(); //cancel current executing threads
+                // Wait a while for tasks to respond to being cancelled
+                if (!executor.awaitTermination(EXECUTOR_SHUTDOWN_TERMINATION_TIMEOUT_MS, TimeUnit.MILLISECONDS))
+                    log.error("Pool did not terminate");
```
Comment (showuon) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: nit: I think we should improve this message a little bit. How about:
`Executor did not terminate in time.`

**Review by showuon — COMMENTED** [review]
Thanks for the update. Looks better. One minor comment left.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java (hunk 2 above)
Comment (karuturi) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Thanks for the review. corrected the message

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the patch. Let's wait for the jenkins build results.

**Comment by karuturi** [comment]
> 
Jenkins has failures unrelated to this PR

**Comment by showuon** [comment]
Failed tests are unrelated:
```
Build / JDK 8 and Scala 2.12 / kafka.network.ConnectionQuotasTest.testListenerConnectionRateLimitWhenActualRateAboveLimit()
```

**Comment by showuon** [comment]
@karuturi , thanks for the contribution!

**Comment by karuturi** [comment]
Thanks for your time @showuon

**Pull request merged**